### PR TITLE
Refactor `Study.optimize`

### PR DESCRIPTION
**Pull request by hvy**
# Motivation

`Study` and `study.py` are becoming quite large and difficult to read and maintain. This PR aims to address the problem by refactoring `Study.optimize`.

# Description 

- Refactors `Study.optimize` by e.g. moving the main logic out to a new file
- Introduces private "ask and tell" interfaces (which could be made public in the future)
- Public API isn't affected

## TODO

- ~I have basically only moved the code out to separate location and would like to go over the necessities of the involved functions and their interfaces~
- ~Ensure unit tests pass~

### CODE REVIEW

**Comment by hvy** [comment]
Ready for review.

This PR might be easier to review commit by commit.

File: optuna/_optimize.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    if not isinstance(catch, tuple):
+        raise TypeError(
+            "The catch argument is of type '{}' but must be a tuple.".format(type(catch).__name__)
+        )
+
+    if not study._optimize_lock.acquire(False):
+        raise RuntimeError("Nested invocation of `Study.optimize` method isn't allowed.")
+
+    # TODO(crcrpar): Make progress bar work when n_jobs != 1.
+    progress_bar = pbar_module._ProgressBar(show_progress_bar and n_jobs == 1, n_trials, timeout)
```
Comment (hvy) [inline optuna/_optimize.py:49]: The progress bar doesn't have to be an attribute/increase the complexity of the `Study`, but can be a local object.

File: optuna/_optimize.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+from optuna import storages
+from optuna import trial as trial_module
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+_logger = logging.get_logger(__name__)
+
+
+def _optimize(
```
Comment (hvy) [inline optuna/_optimize.py:29]: This function is basically a copy of previous logic in `Study.optimize` with some cosmetic fixes such as reduced indirections, reducing arguments being passed around, and handling of the progress bar.

File: optuna/_optimize.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+                break
+            i_trial += 1
+
+        if timeout is not None:
+            elapsed_seconds = (datetime.datetime.now() - time_start).total_seconds()
+            if elapsed_seconds >= timeout:
+                break
+
+        try:
+            trial = _run_trial(study, func, catch)
```
Comment (hvy) [inline optuna/_optimize.py:156]: `_run_trial` no longer cares about `gc`. It's handled here at the caller. Should reduce code complexity.

File: optuna/_optimize.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+        if callbacks is not None:
+            frozen_trial = copy.deepcopy(study._storage.get_trial(trial._trial_id))
+            for callback in callbacks:
+                callback(study, frozen_trial)
+
+        if progress_bar is not None:
+            progress_bar.update((datetime.datetime.now() - time_start).total_seconds())
+
+    study._storage.remove_session()
```
Comment (hvy) [inline optuna/_optimize.py:175]: I believe we should have this method called when exceptions are raised and propagated from `_run_trial` as well. It's sort of orthogonal to this PR.

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your PR. LGTM.

I'm a bit concerned about the file structure. After merging this PR, we can see several files related to the `optuna.study` module or `optuna.study.Study` (i.e., `_study_direction.py`, `_optimize.py` and `study.py`) and they may make the `optuna` directory busy. 
IMHO, we can create `optuna/study` to simplify the `optuna` directory in a follow-up PR.

**Comment by hvy** [comment]
`optuna/study/`, that sounds more than reasonable. Thanks for the review.

**Review by himkt — APPROVED** [review]
Thank you for reducing the complexity of study.py, LGTM!

**Comment by himkt** [comment]
@hvy Sorry, I forgot to merge. Could you please merge a master branch?

**Comment by hvy** [comment]
Sure, done. PTAL!

**Review by himkt — APPROVED** [review]
LGTM again. Thank you for the quick action!

**Comment by hvy** [comment]
Added the milestone since it was missing. Thanks for merging.

**Pull request merged**